src/js/delete.js

```diff
@@ -68,6 +87,62 @@ export function delete_relations(redoing = false) {
   var undo_actions = getUndoActions()
   undo_actions.push(['delete relation', removed.reverse(), selected, extraselected])
   sel.forEach(toggle_selected)
-  if (!redoing)
-    flush_redo()
+  if (!redoing) flush_redo()
+}
+
+/**
+ * Finds all meta-relations that reference a specific relation element, along with their
+ * visual representations and connecting arcs.
+ *
+ * When a relation is deleted, any meta-relations that reference it should also be deleted.
+ * This function identifies all such meta-relations and their associated elements.
+ *
+ * @param {Element} elem - The target relation element being deleted
+ * @param {Document} mei - The MEI document containing the graph data
+ * @param {Array} draw_contexts - The drawing contexts containing SVG representations
+ * @param {Array} svg_hes - Array of SVG elements (this is modified by adding meta-relation SVGs)
+ *
+ * @returns {Object} An object containing:
+ *   - meta_relations: Array of MEI meta-relation nodes that reference the relation
+ *   - meta_relation_arcs: Array of arcs connecting these meta-relations
+ */
+function find_parent_meta_relations(elem, mei, draw_contexts, svg_hes) {
+  let meta_relations = []
+  let meta_relation_arcs = []
+
+  // Find arcs that connect meta-relations to this relation
+  const meta_arcs = Array.from(mei.getElementsByTagName('arc')).filter(
+    arc => arc.getAttribute('to') == '#' + elem.id || arc.getAttribute('from') == '#' + elem.id
+  )
+
+  // For each arc, find the meta-relation nodes
+  meta_arcs.forEach(arc => {
+    // Find the meta-relation (whether it's at the "to" or "from" end)
+    const meta_id =
+      arc.getAttribute('to') === '#' + elem.id
+        ? arc.getAttribute('from').substring(1)
+        : arc.getAttribute('to').substring(1)
+
+    const meta_node = get_by_id(mei, meta_id)
+    if (meta_node && meta_node.getAttribute('type') === 'metarelation') {
+      meta_relations.push(meta_node)
+
+      // Find SVG elements for this meta-relation
+      for (const draw_context of draw_contexts) {
+        let svg_meta = get_by_id(document, draw_context.id_prefix + meta_id)
+        if (svg_meta) {
+          svg_hes.push(svg_meta)
+        }
+      }
+
+      // Find all arcs related to this meta-relation
+      const related_arcs = Array.from(mei.getElementsByTagName('arc')).filter(
+        a => a.getAttribute('to') == '#' + meta_id || a.getAttribute('from') == '#' + meta_id
+      )
+
+      meta_relation_arcs = meta_relation_arcs.concat(related_arcs)
+    }
+  })
+
+  return { meta_relations, meta_relation_arcs }
```

Comment:
Some meta-relations consist of relations. But others consist of meta-relations! There can be, in principle, an arbitrary number of "meta" levels (as demoed on Zoom). So the `find_pattern_meta_relations` method function should somehow work recursively. He can continue the discussion on Mattermost if you want.